flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -412,18 +417,38 @@ private String operatorName(String suffix) {
 
       switch (writeMode) {
         case NONE:
+          if (!equalityFieldIds.isEmpty()) {
+            LOG.info("Distribute rows by equality fields in '{}' distribution mode", DistributionMode.NONE.modeName());
+            return input.keyBy(new EqualityFieldKeySelector(equalityFieldIds, iSchema, flinkRowType));
+          }
+
           return input;
 
         case HASH:
           if (partitionSpec.isUnpartitioned()) {
+            if (!equalityFieldIds.isEmpty()) {
+              LOG.info("Distribute rows by equality fields in '{}' distribution mode, because table is unpartitioned",
+                  DistributionMode.HASH.modeName());
+              return input.keyBy(new EqualityFieldKeySelector(equalityFieldIds, iSchema, flinkRowType));
+            }
+
+            LOG.warn("Fallback to use '{}' distribution mode, because table is unpartitioned",
+                DistributionMode.NONE.modeName());
             return input;
           } else {
+            LOG.info("Distribute rows by partition fields in '{}' distribution mode", DistributionMode.HASH.modeName());
```

Comment:
Why not use both partition and equality field? Why is it okay to ignore equality fields for this case?

Reply:
hash distribution was trying to bring all rows for the same partition value to a single writer. if we include equality field, that would defeat the purpose. On the other hand, if we don't include equality field, we would still suffer the issue that this PR is trying to address. Maybe with equality field, we shouldn't use `HASH` distribution?

Reply:
I discuss the [approach ](https://github.com/apache/iceberg/pull/2898#discussion_r693618065) of combined key with steven before. And I list all cases of distribution in [here](https://github.com/apache/iceberg/pull/2898#discussion_r715845154). 

For this case(table is partitioned and have equality fields and enabled `HASH` distribution), I think we should demand the partition key must be a part of equality key, and we could distribute rows by equality key. Otherwise, we should raise an error and not allow use `HASH` distribution with equality fields.

Reply:
For HASH, we have two cases. First, the hash distribution by partition key may satisfy the required equality distribution. For example, distributing by bucket(16, id) for the equality field id satisfies the requirement because all id values are sent to the same task.

If the partition distribution satisfies the equality distribution, then this should use the partition distribution. Otherwise, it should use the equality distribution to guarantee correctness. It's up to @stevenzwu and you whether you think the latter case should fail and throw an exception ("The required distribution, hash by partition key, does not correctly group the upsert key") or should warn and use the equality distribuiton.

Reply:
When users set HASH distribution, the intention is to cluster data by partition columns. if the hash distribution by partition key doesn't satisfy the required equality distribution, personally I prefer fail with an exception (instead of silently converting the behavior to equality distribution). In this case, I would say it is an incorrect config. We shouldn't set HASH distribution config at all.

Reply:
Strongly agree on loudly, strongly failing (hopefully with a helpful exception) vs silently misbehaving.

Reply:
It seems like a safer way to forbid using HASH distribution with equality fields. I will fix this later.

Reply:
@Reo-LEI, I don't think that's accurate. The requirement is to detect whether the equality distribution is satisfied by the hash distribution. If it is, then you can use the hash distribution. If not, then fail.

I think that check should be whether there are any partition fields that use a source column that is not part of the identifier fields. For example, a table with `type`, `id`, and `data` and a simple key `id` requires any partition transform that uses `id`, like `bucket(256, id)` or `truncate(id, 10)`. But if there is an additional field like `data` then we can't guarantee that all `id` values go to the same data partition.

Another case that does work is when the table identifier is made from both `type` and `id`. In that case, partitioning by either `type` or `id` individually works. Same with partitioning by `type` and `id`. But again, partitioning by data ruins the distribution.

So the logic should be: make sure that the partition spec includes at least one identifier field column and that all of the partition source fields are identifier fields.

Reply:
Thanks @rdblue for the detailed explanation, and @kbendick and @stevenzwu for the advice!

I reorganized the code of distribute. And for this case (table is partitioned and have equality fields and enabled HASH distribution) I added a check to requiring all of the partition source fields are identifier fields. Please take a look again.